File: 005/divisibleByAll.cpp
#include <iostream>
#include <string>
#include <thread>

using namespace std;

typedef unsigned long long superlong;

bool isDivisibleByAll(superlong number, uint start=1, uint end=20)
{
	for(uint i = start; i <= end; i++)
	{
		if(number % i != 0)
			return false;
	}
	return true;
}

void problem5() // brute force
{
	superlong number = 20;
	
	while(true)
	{
		if(isDivisibleByAll(number++))
			break;
	}
	cout << number-1 << " is divisible by each of 1 thru 20!\n";
}

void problem5_better() // brute force optimized
{
	superlong number = 20; // largest amongst divisors
	
	while(true)
	{
		if(isDivisibleByAll(number))
			break;
		number+=20;
	}
	cout << number << " is divisible by each of 1 thru 20!\n";
}


int main()
{	

	problem5_better();

	return 0;
}

File: 005/divisibleByAll.py
#!/usr/bin/env python

def isDivisibleByAll(number):
	for i in xrange(1,21):
		if number % i != 0:
			return False
	return True

def problem5():
	# note you can also set this number below to the lcm of 
	# any of the prior in sequence, for example 1..10 = 2520
	number = 20 # largest number amongst divisors

	while True:
		if isDivisibleByAll(number):
			break
		number += 20
	print "%d is divisible by each of 1 thru 20" % (number)



problem5()